src/kernels/MIOpenSoftmaxAttn.cpp

```diff
@@ -24,33 +24,33 @@
  *
  *******************************************************************************/
 #include <hip/hip_runtime.h>
+//#include <rocrand/rocrand_xorwow.h>
+//#include <rocrand/miopen::prng::xorwow_uniform.h>
 
 #include "miopen_limits.hpp"
 #include "miopen_cstdint.hpp"
 
+#include "miopen_rocrand.hpp"
+
 #ifndef THREADS
 #define THREADS 64
 #endif
 
 namespace {
-template <typename DST, typename SRC>
-constexpr DST bitcast(SRC val)
-{
-    static_assert(sizeof(DST) == sizeof(SRC));
-    DST tmp;
-    /// TODO: wait for C++20 and use std::bit_cast
-    /// until then it's the true way of type puning
-    __builtin_memcpy(&tmp, &val, sizeof(DST));
-    /// unions can work for C99 and later but not for C++,
-    /// though compilers widely and unoficially support it
-    return tmp;
-}
-
 constexpr float max_op(float a, float b) { return a > b ? a : b; };
-constexpr float max_abs_op(float a, float b) { return max_op(abs(a), abs(b)); };
 constexpr float plus_op(float a, float b) { return a + b; };
-constexpr float identety_op(float a) { return a; };
-} // namespace
+
+/// atomically calcutates maximum of non-negative ordered values
+/// it produces wrong results for negatve values or nans
+/// but it is a final amax reducton step and we expect only non-negative ordered values
+__device__ float atomicMaxOfAbsolutValues(float* addr, float value)
+{
+    // ordered non-negatve and even infinity values can be compared as integers
```

Comment:
[R] For IEEE-754-compatible formats, when two floats are compared as UNSIGNED, the following is valid:
```
+0.0 < +denorm < +norm < +INF < +SNAN < +QNAN < -0.0 < -denorm < -norm < -INF < -SNAN < -QNAN
```
So if you use unsigned atomic comparison, then both positive and negative NaNs and INFs are considered greater than positive normal and denormal numbers, which is good for MAX operation. But -0.0 is > than +norm/+denorm, which is wrong.

When two floats are compared as SIGNED:
```
-0.0 < -denorm < -norm < -INF < -SNAN < -QNAN < +0.0 < +denorm < +norm < +INF < +SNAN < +QNAN < 
```
This produces "valid" results with -0.0, but results with -NANs/-INFs will be incorrect (which means that -NANs/-INFs may disappear after atomic max, and that may camouflage correctness issues, if any).

Clearing of MSB guarantees valid result with both -0.0 and -NAN/-INF, but this is impossible to do during atomic max. Therefore I recommend careful analyzing of possible inputs:
- If we are pretty sure that -0.0/-INF/-NAN can't appear on input, then we are fine with both signed and unsigned comparison.
- If there is possibility of -0.0, then I would recommend using SIGNED comparison.
- If only -INF/-NAN may appear, then it is better to use UNSIGNED
- If everything is possible (-0.0/-INF/-NAN) then let's use SIGNED. This guarantees valid results with -0.0, but who really cares about disappearing -INF/-NAN? Seriously, if we have -INF/-NAN then it is likely that we also have +INF/+NAN, which do not disappear.

It's up to you what to do with the above considerations. I guess that at least you may wish to add some comments to the code.

Reply:
I'm aware about this behavior. This function must be called only for non-negative-ordered numbers.

It does not work with NANs, since `fmax(val, NAN) = val` (because all the comparisons `val` vs `NAN` if always `false`), but this function returns a `NAN`, and it does not work with the negative values, which has already been stated in the comment.
C++ std: https://en.cppreference.com/w/cpp/numeric/math/fmax
CUDA: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1g6e7516db46be25c33fb26e203287f2a3

As for the input data - prior to that atomic operation we do `fmax` operation with initial `0` and even if there is any `NAN` in the input data, it will be filtered and replaced by 0 or any other positive value.

Reply:
@CAHEK7 
> This function must be called only for non-negative-ordered numbers.

Yes, but the question is: can we provide a guarantee that this function will be called with such a numbers?

If not, then another question: Do we want NaNs to disappear or to remain? The answer depend on what is required from the softmax primitive.

> As for the input data - prior to that atomic operation we do `fmax` operation with initial `0` and even if there is any `NAN` in the input data, it will be filtered and replaced by 0 or any other positive value.

Is it what we really want?

Reply:
> Yes, but the question is: can we provide a guarantee that this function will be called with such a numbers?
Yes, because the value prior this atomic operation acquired from `fmax(0, val)` which filters all the `NANs`

> Is it what we really want?
Kind of, the only case where it can produce different result against normal `fmax` reduction is the buffer of all `NANs` - we will produce 0. In the case where we have at least 1 normal number we will match.

The documentation does not say anything explicit about that case.

Reply:
@CAHEK7 Let me summarize: **_The primitive does not guarantee correct (well-defined) results when input is invalid (i.e. contains INFs or NANs)._** Correct behavior in this case means that if input is invalid then output should be invalid as well. Invalid output means INF or NAN; it is questionable that output 0 should be considered invalid (it is meaningless, but formally valid).

Please unhide this comment.

[resolved]